### PR TITLE
Fix NFT > CAT Royalty splitting bug.

### LINKED ISSUE #12359: Trying to accept an NFT offer for CAT throws error unless you first send 500 CATs to yourself
### What happened?

This issue comes by way of a Twitter thread where user `@kimsk` was trying to purchase an NFT from `@endertown` for 1,000 `CATMOS`, but was getting an Error in the GUI when clicking to accept. 
https://twitter.com/hoffmang/status/1546677026081292288?s=20&t=u50T5xr17RARiZHgcSaAcg

@hoffmang9 replied implying that he too was having the issue with `@endertown`'s offers, and apparently the workaround was that the person buying needed to first send 500 `CATMOS` to themselves in order for the offer to be accepted.

### Version

1.4.0

### What platform are you using?

Windows

### What ui mode are you using?

GUI

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by jack60612**
Add comments, and move some things around to increase readability
+ fix failure on extra cat delta from arbitrage

### CODE REVIEW

**Comment by jack60612** [comment]
Did not work before change, after changes were applied Offer was able to be executed.

File: chia/wallet/trading/offer.py [hunk 1]
```diff
@@ -198,6 +198,11 @@ def get_requested_amounts(self) -> Dict[Optional[bytes32], int]:
         return requested_amounts
 
     def arbitrage(self) -> Dict[Optional[bytes32], int]:
+        """
+        Returns a dictionary of the amount and type of each asset that is involved in the trade
+        If we are requesting an asset of Type A, then the amount of that asset will be negative
+        However, if we are offering an asset of Type B, then the amount of that asset (int) will be positive.
+        """
```
Comment (Quexington) [inline chia/wallet/trading/offer.py:205]: This comment is not correct.  This method has no idea who the party running it is so "we" cannot be a concept.  This method returns a dictionary of assets and how much their offered amount within the offer exceeds/falls short of their requested amount.

**Review by Quexington — CHANGES_REQUESTED** [review]

File: chia/wallet/trade_manager.py [hunk 2]
```diff
@@ -498,9 +500,13 @@ async def check_offer_validity(self, offer: Offer) -> bool:
...
 
-    async def calculate_tx_records_for_offer(self, offer: Offer, validate: bool) -> List[TransactionRecord]:
+    async def calculate_tx_records_for_offer(
+        self, offer: Offer, validate: bool, arbitrage_ph: Optional[bytes32] = None
+    ) -> List[TransactionRecord]:
         if validate:
-            final_spend_bundle: SpendBundle = offer.to_valid_spend()
+            final_spend_bundle: SpendBundle = offer.to_valid_spend(
+                arbitrage_ph
+            )  # arbitrage_ph is where the extra CAT's go.
```
[the remark this answers is not in the document]
Comment (jack60612) [reply chia/wallet/trade_manager.py:509]: If there is cat change then it goes to this wallet, that was the 2nd issue causing this

[the remark this answers is not in the document]
Comment (Quexington) [reply chia/wallet/trade_manager.py:509]: The issue was that the change coin and the royalty coin has the same amount, there should not be any arbitrage in the offer after the first fix.  I haven't tested this, but that's my assumption.

[the remark this answers is not in the document]
Comment (jack60612) [reply chia/wallet/trade_manager.py:509]: > The issue was that the change coin and the royalty coin has the same amount, there should not be any arbitrage in the offer after the first fix. I haven't tested this, but that's my assumption.
It failed even after fixing that, and this fix was also needed.

File: chia/wallet/trading/offer.py (hunk 1 above)
Comment (jack60612) [reply chia/wallet/trading/offer.py:205]: Yeah im bad at wording things, let me fix it.

File: chia/wallet/trade_manager.py [hunk 3]
```diff
@@ -625,12 +631,16 @@ async def respond_to_offer(
...
         complete_offer = Offer.aggregate([offer, take_offer])
+        self.log.info(f"COMPLETE OFFER: {complete_offer.to_bech32()}")
         assert complete_offer.is_valid()
-
-        final_spend_bundle: SpendBundle = complete_offer.to_valid_spend()
+        wallet_ph: Optional[bytes32] = None
+        if sum(complete_offer.arbitrage().values()) > 0:
+            # we need an address in case we end up with extra coins.
+            wallet_ph = await self.wallet_state_manager.main_wallet.get_puzzle_hash(False)
+        final_spend_bundle: SpendBundle = complete_offer.to_valid_spend(wallet_ph)
```
[the remark this answers is not in the document]
Comment (jack60612) [reply chia/wallet/trade_manager.py]: added a proper fix, that took far too long.

File: chia/wallet/trade_manager.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jack60612) [reply chia/wallet/trade_manager.py:509]: added correct fix

File: chia/wallet/nft_wallet/nft_wallet.py [hunk 4]
```diff
@@ -810,34 +811,33 @@ async def make_nft1_offer(
...
+                    offered_asset_id: [
+                        NotarizedPayment(royalty_address, royalty_amount, [royalty_address], requested_asset_id)
+                    ]
+                },
+                driver_dict,
             )
-
+            # add royalty announcement to the list of announcements to assert
+            announcements_to_assert.extend(royalty_announcement)
+            royalty_ph = royalty_announcement[0].origin_info
```
Comment (Quexington) [inline chia/wallet/nft_wallet/nft_wallet.py:840]: I wrote this in a comment before some pushes but I don't like using the puzhash from one of these calculated announcements being used to set `royalty_ph`.  We should instead be constructing the puzzle hash ourselves `construct_puzzle`, the driver_dict, and `OFFER_MOD`.

File: chia/wallet/trade_manager.py (hunk 2 above)
Comment (Quexington) [inline chia/wallet/trade_manager.py:509]: Can we remove this bit since we don't support arbitrage right now?

File: chia/wallet/trading/offer.py [hunk 5]
```diff
@@ -155,22 +155,23 @@ def get_offered_coins(self) -> Dict[Optional[bytes32], List[Coin]]:
                 inner_solution: Optional[Program] = get_inner_solution(puzzle_driver, parent_solution)
                 assert inner_puzzle is not None and inner_solution is not None
                 conditions: Program = inner_puzzle.run(inner_solution)
+                spend_additions: List[Coin] = []
```
Comment (Quexington) [inline chia/wallet/trading/offer.py:158]: This is nitpicky, but I don't like this rename.  It's specifically the additions with the amount matching the condition that creates the OFFER_HASH.

File: chia/wallet/trade_manager.py (hunk 2 above)
Comment (jack60612) [reply chia/wallet/trade_manager.py:509]: yep!

File: chia/wallet/nft_wallet/nft_wallet.py (hunk 4 above)
Comment (jack60612) [reply chia/wallet/nft_wallet/nft_wallet.py:840]: ah i misunderstood what you meant, gotcha!

**Review by Quexington — APPROVED** [review]

**Review by trepca — APPROVED** [review]

**Pull request merged**